### PR TITLE
[exporterhelper] New exporter helper for custom requests

### PR DESCRIPTION
**Pull request by dmitryax**
Introduce a new exporter helper that operates over client-provided requests instead of pdata. The helper user now have to provide:
- Converter: an interface with a function implementing translation of pdata Metrics/Traces/Logs into a user-defined Request.
- Request sender: a function that will be called to send the Request

It opens a door for moving batching to the exporter, where batches will be built from client data format, instead of pdata. The batches can be properly sized by custom request size, which can be different from OTLP. The same custom request sizing will be applied to the sending queue. It will also improve performance of the sending queue retries for non-OTLP exporters, they don't need to translate pdata on every retry. 

This is an experimental API. Once stabilized, it's intended to replace the existing helpers.

Tracking Issue: https://github.com/open-telemetry/opentelemetry-collector/issues/8122

Related issue: https://github.com/open-telemetry/opentelemetry-collector/issues/4646

### CODE REVIEW

**Review by jmacd — APPROVED** [review]
Looking forward! I see this as an important step for batching by request size.

**Comment by dmitryax** [comment]
@open-telemetry/collector-approvers, please review whenever you have a chance. I'm planning to put more PR's on top of it next week

File: exporter/exporterhelper/common.go
Comment (mx-psi) [inline exporter/exporterhelper/common.go:137]: I would at least raise a warning about this, possibly just error out

File: exporter/exporterhelper/request.go
Comment (mx-psi) [inline exporter/exporterhelper/request.go:14]:
```suggestion
	// ItemsCount returns the count basic item in the request, the smallest pieces of data that can be sent to the endpoint.
```

File: exporter/exporterhelper/request.go
Comment (mx-psi) [inline exporter/exporterhelper/request.go:13]: Should we make this interface unimplementable and have some public base request struct to provide default implementations for future methods?

File: exporter/exporterhelper/logs.go
Comment (mx-psi) [inline exporter/exporterhelper/logs.go:131]: nit: We could name this `NewLogsRequestExporter` instead of using the `V2` suffix

**Review by mx-psi — COMMENTED** [review]
Apologies for the delay in reviewing this, I missed this

File: exporter/exporterhelper/logs.go
Comment (bogdandrutu) [reply exporter/exporterhelper/logs.go:131]: We have not stabilized the previous API, so I agree with @mx-psi on this.

File: exporter/exporterhelper/logs.go
Comment (bogdandrutu) [inline exporter/exporterhelper/logs.go:64]: Should you delete this func?

File: exporter/exporterhelper/logs.go
Comment (bogdandrutu) [reply exporter/exporterhelper/logs.go:131]: Also would be good if we can eventually make old func call into the new.

File: exporter/exporterhelper/request.go
Comment (bogdandrutu) [reply exporter/exporterhelper/request.go:13]: Alternative is to have also `RequestItemer` (or something like that) so we just don't need an interface and we use `any`. And ask users to provide funcs that work with the request. Another option is to ask user to implement the Send on that request as well, but right now `ItemCount` is on the request but Send is not, so inconsistent.

Comment (dmitryax) [reply exporter/exporterhelper/request.go:13]: > Alternative is to have also RequestItemer (or something like that) so we just don't need an interface and we use `any`. And ask users to provide funcs that work with the request

In that case, do you suggest we make this an optional interface? We could do that, but the new itemized queue and batching exporter helpers won't work. We can probably allow that in case clients don't need that.

> Another option is to ask user to implement the Send on that request as well, but right now ItemCount is on the request but Send is not, so inconsistent.

What if we ask clients to implement `Request` with only one required `Send` function, and if they want itemized queue or batching helpers, they implement the optional interface `RequestItemer`. Then, we can possibly introduce bytes-sized queue or batching helpers if they implement `RequestSizer`.

```golang
type Request interface {
	Send(ctx context.Context) error
}

type RequestItemer interface {
	ItemsCount()  int
}

type RequestSizer interface {
	BytesSize() int
}
```

WDYT?

File: exporter/exporterhelper/common.go
Comment (dmitryax) [reply exporter/exporterhelper/common.go:137]: The new exporter helper will get another option for the persistent queue with marshaller and unmarshaller as required params, and maybe another option for in-memory queue when the config doesn't have the `storage` param. So we won't be able to reuse this option anyway. I'll just put a panic here if it's used with the new exporter helper.

File: exporter/exporterhelper/logs.go
Comment (dmitryax) [reply exporter/exporterhelper/logs.go]: SGTM. Updated

File: exporter/exporterhelper/logs.go
Comment (dmitryax) [reply exporter/exporterhelper/logs.go]: Right. It's not needed anymore. Removed

File: exporter/exporterhelper/logs.go
Comment (dmitryax) [reply exporter/exporterhelper/logs.go]: > Also would be good if we can eventually make old func call into the new.

I didn't add queue options for the new helper in this PR yet. Wanted to do it as the next step. Once it's done, we can make the old func call the new one.

**Comment by jpkrohling** [comment]
I would like to review this before it gets merged, especially if it conflicts with https://github.com/open-telemetry/opentelemetry-collector/pull/7510.

File: exporter/exporterhelper/request.go
Comment (dmitryax) [reply exporter/exporterhelper/request.go:13]: I submitted another PR with the suggested approach

**Pull request closed without merging**
